colorama/ansitowin32.py

```diff
@@ -87,14 +87,21 @@ def __init__(self, wrapped, convert=None, strip=None, autoreset=False):
         # to support the ANSI codes.
         conversion_supported = on_windows and winapi_test()
 
+        # check if this is a TTY or if the FORCE_COLOR environment variable
+        # has been set and NO_COLOR has not been set.
+        not_closed = not self.stream.closed
+        force_color = 'FORCE_COLOR' in os.environ.keys()
+        no_color = 'NO_COLOR' in os.environ.keys()
+        color_allowed = ((not_closed and self.stream.isatty()) or force_color) and not no_color
```

Comment:
I may be wrong, but does `not_closed` have meaning here? We check for `not_closed` later, for `strip` and `convert` again, and I think that the expressions there mean we can skip the check for `not_closed` here.
(if we find out we only need to check it in one place - I prefer to leave the full expression, `not self.stream.closed`, because negative booleans, i.e. any `not_*` variables, are kinda confusing IMO)

Reply:
Yes - it is rather confusing, but in terms of the logical structure, it appears to be optimal, in that I optimized it using Yosys (see above).

Reply:
I'm not familiar with Yosys, and it appears that it didn't change the original logic. From my "pen and paper" calculations (and also a test with a small Python script) I still think that removing `not closed` from the `color_allowed` expression should not affect the result or `convert` and `strip` regardless of the values of the inputs. Maybe Yosys only optimizes logic if there are more complex changes like removing inputs or something like that? Or I missed something?

Reply:
Ok... I've just pushed a branch with the following logic:

```
        not_closed = not self.stream.closed
        force_color = 'FORCE_COLOR' in os.environ.keys()
        no_color = 'NO_COLOR' in os.environ.keys()
        color_allowed = (self.stream.isatty() or force_color) and not no_color
```

but it results in [test errors](https://travis-ci.org/tartley/colorama/builds/657729965?utm_source=github_status&utm_medium=notification), because you can't do `isatty` on a closed stream! So I think the logic has to stay as-is - though I think everyone who's looked at this code found it confusing.

So I decided to break it up with more comments.